pmlc/conversion/tile_to_pxa/specials.cc

```diff
@@ -675,6 +702,39 @@ struct GatherOpConversion : public OpConversionPattern<tile::GatherOp> {
     auto valuePlusHalf = rewriter.create<mlir::AddFOp>(loc, value, half);
     return floorFPToSI(loc, rewriter, valuePlusHalf, integerType);
   }
+
+  Value processPaddingMode(Location loc, ConversionPatternRewriter &rewriter,
+                           Value result, Value idx, IndexBounds bounds,
+                           Type elementType, PaddingMode paddingMode) const {
+    switch (paddingMode) {
+    case PaddingMode::edge:
+      break;
```

Comment:
How does this accomplish edge padding? Does interpolation work for gather with an out-of-bounds read by getting the closest value (i.e. at the edge)? That's not what I had thought but perhaps I'm mistaken.

Reply:
In previous version, the closest value will be read when the index is out-of-bounds, So it doesn't need any change when PaddingMode is edge. But in some operations, for example deformabel_convolution (https://github.com/plaidml/openvino/blob/plaidml/ngraph/core/reference/include/ngraph/runtime/reference/deformable_convolution.hpp, lines 157 to 160) zero will be read when the index is out-of-bounds. So, i added the zero_padding_mode to process this case.

Reply:
Gather with an out-of-bounds read by getting the closest value (i.e. at the edge) is true for the linear interpolation (not true for cubic or nearest mode). The `paddingMode::edge` was the default behavior for interpolated gather (the same as ngraph ref implementation of interpolation). That is when the index is out-of-bounds, the data will be treated as if it was edge padded. 

For example, for linear interpolated gather with index at -0.3 for a 1D data with 2 element {data[0]=3; data[1]=4;}, the data will be treated as if the data[-1]=data[0]=3, so any linear interpolation  between -1 and 0 will be 3 (It looks like it gathers the nearest edge, but it is actually calculated by data[-0.3]=data[-1]*0.7+data[0]*0.3=3).

For cubic interpolated gather with index at -0.3 for a 1D data with 4 element {data[0]=3; data[1]=4;data[2]=3; data[3]=2;}
the result of data[-0.3] will be calculated by using four points, data[-1],data[0],data[1],data[2], since data[-1] is out-of-bounds, it will be set to data[-1]=data[0]=3, then use these four values to calculate data[-0.3] and it does not equal the edge value data[0].

As Ding mentioned, some ngraph ref implementation like deformableconvolution uses different out-of-bounds values than the interpolation op. They return 0 when the index is out of bounds. Ding added this function to return zero for linear interpolation when the index is out of bounds. I have applied it to all interpolation mode.

I think it is a little confusing to use the word `PaddingMode`, since in `PaddingMode::edge` we treat the data as if it was edge padded, in `PaddingMode::zero` we simply return zero when index is out of bounds (This is a different then treating the data as if it was zero padded). I think a better word for this mode would be `OutOfBoundsMode::data_edge_padding`  `OutOfBoundsMode::return_zero`